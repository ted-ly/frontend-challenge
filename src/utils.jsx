import { useState, useCallback } from 'react';

export const initialState = {
    name: '',
    password: '',
    color: '',
    email: '',
    terms: false,
};

export const onClickNext = (to, setPathname) => {
    setPathname(to);
    window.history.pushState(null, '', to);
};

export const onClickBack = (from, setPathname) => {
    setPathname(from);
    window.history.pushState(null, '', from);
}; 

export const onSubmit = async (formData, setPathname) => {
    const resp = await fetch('http://localhost:3001/api/submit', { method: 'POST', headers: {'Content-Type': 'application/json'}, body: JSON.stringify(formData) });
    if (resp.status >= 400) {
        window.history.pushState(null, '', '/error');
        setPathname('/error');
    } else {
        window.history.pushState(null, '', '/success');
        setPathname('/success');
    }
};

export const createOnClickRestart = ({ setFormData, setPathname }) => useCallback(() => {
    window.history.pushState(null, '', '/');
    setPathname('/');
    setFormData({...initialState});
}, []);

export const createOnFormDataChange = ({ setFormData, formData }) => useCallback((e) => {
    const name = e.target.name;
    const value = e.target.value;
    const newFormData = {
        ...formData,
        [name]: name ==='terms' ? !formData.terms : value,
    };

    setFormData({...newFormData});
}, [formData]); 

export const useForm = () => {
    const [formData, setFormData] = useState(initialState);
    const [pathname, setPathname] = useState('/');
    const onFormDataChange = createOnFormDataChange({ formData, setFormData });
    const onClickRestart = createOnClickRestart({ setFormData, setPathname });

    return { onClickRestart, onFormDataChange, formData, setPathname };
};

export const toCapitalized = (str) => {
    return str.slice(0, 1).toUpperCase() + str.slice(1);
}

export const toTransformText = (text, key) => {
    if (key === 'terms') {
        return text ? 'Agreed' : 'Disagreed';
    }

    if (key === 'password') {
        return '*'.repeat(text.length);
    }

    return text;
}; 