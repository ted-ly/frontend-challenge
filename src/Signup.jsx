import React, { useCallback, useMemo, useEffect } from 'react';
import Input from './Input';
import Step from './Step';
import Button from './Button';
import { onClickNext } from './utils';

const textKeys = {
    name: {placeholder: 'First Name', type: 'text'},
    password: {placeholder: 'Password', type: 'password'},
    email: {placeholder: 'E-mail', type: 'email'},
};
export default function Signup({ formData, onFormDataChange, setPathname }) {
    const inputFields = useMemo(() => {
        return Object.keys(textKeys).map(key => <Input type={textKeys[key].type} key={key} placeholder={textKeys[key].placeholder} name={key} onChange={onFormDataChange} value={formData[key]}/>);
    }, [formData.name, formData.password, formData.email]);

    const onClick = useCallback(() => {
        onClickNext('/more-info', setPathname);
    }, []);
    return (
        <Step title='Sign Up'>
            <div className='signup-container'>
                {inputFields}
                <div className='button-container'>
                    <Button onClick={onClick}>Next</Button>
                </div>
            </div>
        </Step>
    )
}
