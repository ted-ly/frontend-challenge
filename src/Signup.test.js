import renderer from 'react-test-renderer';
import Signup from './Signup';
import { initialState } from './utils';

describe('Signup', () => {
    it('should match snapshot', () => {
        const component = renderer.create(<Signup formData={initialState} />);
        expect(component.toJSON()).toMatchSnapshot();
    });
});