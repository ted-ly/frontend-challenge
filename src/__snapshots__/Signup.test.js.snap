// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup should match snapshot 1`] = `
<div
  className="step-container"
>
  <div
    className="step-title"
  >
    Sign Up
  </div>
  <div
    className="signup-container"
  >
    <input
      name="name"
      placeholder="First Name"
      type="text"
      value=""
    />
    <input
      name="password"
      placeholder="Password"
      type="password"
      value=""
    />
    <input
      name="email"
      placeholder="E-mail"
      type="email"
      value=""
    />
    <div
      className="button-container"
    >
      <button
        onClick={[Function]}
        type="button"
      >
        Next
      </button>
    </div>
  </div>
</div>
`;
